code/modules/projectiles/guns/energy/special.dm

```diff
@@ -16,6 +16,45 @@
 /obj/item/weapon/gun/energy/ionrifle/emp_act(severity)
 	return
 
+/obj/item/weapon/gun/energy/ionrifle/attackby(var/obj/item/A as obj, mob/user as mob)
+	..()
+	A.update_icon()
+	update_icon()
+	if(istype(A, /obj/item/tool/circular_saw) || istype(A, /obj/item/weapon/melee/energy) || istype(A, /obj/item/weapon/pickaxe/plasmacutter))
+		to_chat(user, "<span class='notice'>You begin to shorten the barrel of \the [src].</span>")
+		if(prob(25))
+			afterattack(user, user)	//will this work?
+			afterattack(user, user)	//it will. we call it twice, for twice the FUN
+			playsound(user, fire_sound, 50, 1)
+			user.visible_message("<span class='danger'>The [src] goes off!</span>", "<span class='danger'>The [src] goes off in your face!</span>")
+			return
```

Comment:
just make this a fire_inhand() proc if you're going to paste it, more maintainable and it can be used for other things